File: Sources/LoafCat/PixelCanvas.swift
import AppKit

/// An RGBA colour in 8-bit channels, straight from the atlas palette.
struct RGBA {
    var r: UInt8, g: UInt8, b: UInt8, a: UInt8

    static let clear = RGBA(r: 0, g: 0, b: 0, a: 0)

    /// Parses `#RRGGBBAA` (or `#RRGGBB`). The atlas writes the 8-digit form.
    init?(hex: String) {
        var s = Substring(hex)
        if s.hasPrefix("#") { s = s.dropFirst() }
        guard s.count == 6 || s.count == 8, let v = UInt32(s, radix: 16) else { return nil }
        if s.count == 6 {
            r = UInt8((v >> 16) & 0xFF); g = UInt8((v >> 8) & 0xFF)
            b = UInt8(v & 0xFF); a = 255
        } else {
            r = UInt8((v >> 24) & 0xFF); g = UInt8((v >> 16) & 0xFF)
            b = UInt8((v >> 8) & 0xFF); a = UInt8(v & 0xFF)
        }
    }

    init(r: UInt8, g: UInt8, b: UInt8, a: UInt8) {
        self.r = r; self.g = g; self.b = b; self.a = a
    }

    var cgColor: CGColor {
        CGColor(red: CGFloat(r) / 255, green: CGFloat(g) / 255,
                blue: CGFloat(b) / 255, alpha: CGFloat(a) / 255)
    }
}

/// A surface of whole LOGICAL pixels, composed at 1x and magnified by an integer
/// factor with nearest-neighbour at draw time.
///
/// Composing UI art at 1x and magnifying is the only way to keep it in the same
/// visual language as the cat. Anything drawn at device resolution — a
/// `CGContext` at 2x, a `CATextLayer`, an `NSAttributedString` — lands on
/// half-pixels and anti-aliases its own edges, and no amount of downstream care
/// gets that back.
struct PixelBitmap {
    private(set) var width: Int
    private(set) var height: Int
    /// Row-major RGBA, row 0 at the TOP, matching atlas coordinates.
    private(set) var rgba: [UInt8]

    init(width: Int, height: Int) {
        self.width = max(width, 0)
        self.height = max(height, 0)
        self.rgba = [UInt8](repeating: 0, count: self.width * self.height * 4)
    }

    /// Reads a PNG into logical pixels. Only ever called at load time.
    init?(contentsOf url: URL) {
        guard
            let src = CGImageSourceCreateWithURL(url as CFURL, nil),
            let img = CGImageSourceCreateImageAtIndex(src, 0, nil)
        else { return nil }
        self.init(cgImage: img)
    }

    init?(cgImage img: CGImage) {
        let w = img.width, h = img.height
        guard w > 0, h > 0 else { return nil }
        var buf = [UInt8](repeating: 0, count: w * h * 4)
        let ok: Bool = buf.withUnsafeMutableBytes { raw -> Bool in
            guard let ctx = CGContext(
                data: raw.baseAddress, width: w, height: h, bitsPerComponent: 8,
                bytesPerRow: w * 4, space: CGColorSpaceCreateDeviceRGB(),
                bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue)
            else { return false }
            ctx.interpolationQuality = .none
            ctx.draw(img, in: CGRect(x: 0, y: 0, width: w, height: h))
            return true
        }
        guard ok else { return nil }

        // No flip. A CGBitmapContext's USER SPACE is y-up, but its backing store is
        // top-down: buffer row 0 is the image's top row, which is already what the
        // atlas means by y. Measured with an asymmetric part (see spikes/hitmask),
        // because getting this backwards silently mirrors every glyph.
        self.width = w
        self.height = h
        self.rgba = buf
    }

    subscript(x: Int, y: Int) -> RGBA {
        get {
            guard x >= 0, x < width, y >= 0, y < height else { return .clear }
            let i = (y * width + x) * 4
            return RGBA(r: rgba[i], g: rgba[i + 1], b: rgba[i + 2], a: rgba[i + 3])
        }
        set {
            guard x >= 0, x < width, y >= 0, y < height else { return }
            let i = (y * width + x) * 4
            rgba[i] = newValue.r; rgba[i + 1] = newValue.g
            rgba[i + 2] = newValue.b; rgba[i + 3] = newValue.a
        }
    }

    /// Copies every non-transparent pixel of `src`. Opaque-or-clear art only, so a
    /// straight copy is the correct blend and there is nothing to round.
    mutating func blit(_ src: PixelBitmap, x: Int, y: Int) {
        for sy in 0..<src.height {
            for sx in 0..<src.width {
                let c = src[sx, sy]
                if c.a > 0 { self[x + sx, y + sy] = c }
            }
        }
    }

    /// Inks a sub-rectangle of a coverage mask (the font sheet) in a flat colour.
    mutating func stamp(mask: PixelBitmap, from: CGRect, x: Int, y: Int, color: RGBA) {
        let mx = Int(from.origin.x), my = Int(from.origin.y)
        for sy in 0..<Int(from.height) {
            for sx in 0..<Int(from.width) where mask[mx + sx, my + sy].a > 128 {
                self[x + sx, y + sy] = color
            }
        }
    }

    func cgImage() -> CGImage? {
        guard width > 0, height > 0,
              let provider = CGDataProvider(data: Data(rgba) as CFData) else { return nil }
        return CGImage(
            width: width, height: height, bitsPerComponent: 8, bitsPerPixel: 32,
            bytesPerRow: width * 4, space: CGColorSpaceCreateDeviceRGB(),
            bitmapInfo: CGBitmapInfo(rawValue: CGImageAlphaInfo.premultipliedLast.rawValue),
            provider: provider, decode: nil, shouldInterpolate: false,
            intent: .defaultIntent)
    }
}

File: windows/LoafCat/PixelCanvas.cs
using System.Drawing;
using System.Drawing.Imaging;
using System.Globalization;
using System.Runtime.Versioning;

namespace LoafCat;

/// An RGBA colour in 8-bit channels, straight from the atlas palette.
public struct Rgba(byte r, byte g, byte b, byte a)
{
    public byte R = r, G = g, B = b, A = a;

    public static readonly Rgba Clear = new(0, 0, 0, 0);

    /// Parses `#RRGGBBAA` (or `#RRGGBB`). The atlas writes the 8-digit form.
    public static Rgba? FromHex(string? hex)
    {
        if (string.IsNullOrEmpty(hex)) return null;
        ReadOnlySpan<char> s = hex.AsSpan();
        if (s.Length > 0 && s[0] == '#') s = s[1..];
        if (s.Length != 6 && s.Length != 8) return null;
        if (!uint.TryParse(s, NumberStyles.HexNumber, CultureInfo.InvariantCulture, out uint v))
            return null;

        return s.Length == 6
            ? new Rgba((byte)(v >> 16), (byte)(v >> 8), (byte)v, 255)
            : new Rgba((byte)(v >> 24), (byte)(v >> 16), (byte)(v >> 8), (byte)v);
    }
}

/// A surface of whole LOGICAL pixels, composed at 1x and magnified by an integer
/// factor with nearest-neighbour at draw time.
///
/// Composing UI art at 1x and magnifying is the only way to keep it in the same
/// visual language as the cat. Anything drawn at device resolution — a `Graphics`
/// scaled by 2, a `DrawString`, a WinForms label — lands on half-pixels and
/// anti-aliases its own edges, and no amount of downstream care gets that back.
///
/// Straight (non-premultiplied) RGBA, row 0 at the TOP, matching both the atlas and
/// the macOS port byte for byte. Premultiplication happens once, in the compositor,
/// because that is the only place that needs it.
public sealed class PixelBitmap
{
    public int Width { get; }
    public int Height { get; }

    /// Row-major RGBA. Public because the compositor's inner loop indexes it
    /// directly and a property accessor per pixel at 120Hz is not free.
    public readonly byte[] Pixels;

    public PixelBitmap(int width, int height)
    {
        Width = Math.Max(width, 0);
        Height = Math.Max(height, 0);
        Pixels = new byte[Width * Height * 4];
    }

    private PixelBitmap(int width, int height, byte[] rgba)
    {
        Width = width;
        Height = height;
        Pixels = rgba;
    }

    /// Reads a PNG into logical pixels. Only ever called at load time.
    ///
    /// GDI+ hands back BGRA in memory for Format32bppArgb (little-endian "ARGB"),
    /// so the channel swap below is a fact about the platform rather than a choice.
    /// Getting it backwards is invisible on the mono theme, whose coat is grey —
    /// which is exactly how a bug like that survives to the first colour theme.
    [SupportedOSPlatform("windows")]
    public static PixelBitmap? Load(string path)
    {
        if (!File.Exists(path)) return null;
        try
        {
            // Loaded through a MemoryStream, not Image.FromFile: that keeps a lock on
            // the file for the lifetime of the Bitmap, and the theme directory has to
            // stay replaceable while the app is running.
            byte[] raw = File.ReadAllBytes(path);
            using var ms = new MemoryStream(raw, writable: false);
            using var src = new Bitmap(ms);
            return FromBitmap(src);
        }
        catch (Exception)
        {
            return null;
        }
    }

    [SupportedOSPlatform("windows")]
    public static PixelBitmap? FromBitmap(Bitmap src)
    {
        int w = src.Width, h = src.Height;
        if (w <= 0 || h <= 0) return null;

        var buf = new byte[w * h * 4];
        // Clone into a known format first: the source may be paletted, 24-bit, or
        // carry a colour profile, and LockBits will convert only if asked to.
        using var norm = src.Clone(new Rectangle(0, 0, w, h), PixelFormat.Format32bppArgb);
        var data = norm.LockBits(
            new Rectangle(0, 0, w, h), ImageLockMode.ReadOnly, PixelFormat.Format32bppArgb);
        try
        {
            unsafe
            {
                byte* base_ = (byte*)data.Scan0;
                for (int y = 0; y < h; y++)
                {
                    byte* row = base_ + (long)y * data.Stride;
                    int o = y * w * 4;
                    for (int x = 0; x < w; x++)
                    {
                        buf[o + 0] = row[x * 4 + 2];   // R <- B
                        buf[o + 1] = row[x * 4 + 1];   // G
                        buf[o + 2] = row[x * 4 + 0];   // B <- R
                        buf[o + 3] = row[x * 4 + 3];   // A
                        o += 4;
                    }
                }
            }
        }
        finally
        {
            norm.UnlockBits(data);
        }
        return new PixelBitmap(w, h, buf);
    }

    public Rgba this[int x, int y]
    {
        get
        {
            if ((uint)x >= (uint)Width || (uint)y >= (uint)Height) return Rgba.Clear;
            int i = (y * Width + x) * 4;
            return new Rgba(Pixels[i], Pixels[i + 1], Pixels[i + 2], Pixels[i + 3]);
        }
        set
        {
            if ((uint)x >= (uint)Width || (uint)y >= (uint)Height) return;
            int i = (y * Width + x) * 4;
            Pixels[i] = value.R;
            Pixels[i + 1] = value.G;
            Pixels[i + 2] = value.B;
            Pixels[i + 3] = value.A;
        }
    }

    /// Copies every non-transparent pixel of `src`. Opaque-or-clear art only, so a
    /// straight copy is the correct blend and there is nothing to round.
    public void Blit(PixelBitmap src, int x, int y)
    {
        for (int sy = 0; sy < src.Height; sy++)
        {
            for (int sx = 0; sx < src.Width; sx++)
            {
                var c = src[sx, sy];
                if (c.A > 0) this[x + sx, y + sy] = c;
            }
        }
    }

    /// Inks a sub-rectangle of a coverage mask (the font sheet) in a flat colour.
    public void Stamp(PixelBitmap mask, int fromX, int fromY, int w, int h,
                      int x, int y, Rgba color)
    {
        for (int sy = 0; sy < h; sy++)
        {
            for (int sx = 0; sx < w; sx++)
            {
                if (mask[fromX + sx, fromY + sy].A > 128) this[x + sx, y + sy] = color;
            }
        }
    }

    /// True when nothing in here would draw. Used to skip empty overlay slots.
    public bool IsBlank()
    {
        for (int i = 3; i < Pixels.Length; i += 4)
        {
            if (Pixels[i] != 0) return false;
        }
        return true;
    }

    /// A GDI+ bitmap of the same pixels, magnified by an integer factor with
    /// nearest-neighbour. Only used for chrome — the tray icon and the theme
    /// thumbnails in Settings. The cat itself never goes through GDI+; see CatView.
    [SupportedOSPlatform("windows")]
    public Bitmap ToBitmap(int magnify = 1)
    {
        magnify = Math.Max(magnify, 1);
        var outBmp = new Bitmap(Width * magnify, Height * magnify, PixelFormat.Format32bppArgb);
        var data = outBmp.LockBits(
            new Rectangle(0, 0, outBmp.Width, outBmp.Height),
            ImageLockMode.WriteOnly, PixelFormat.Format32bppArgb);
        try
        {
            unsafe
            {
                byte* base_ = (byte*)data.Scan0;
                for (int y = 0; y < outBmp.Height; y++)
                {
                    byte* row = base_ + (long)y * data.Stride;
                    int sy = y / magnify;
                    for (int x = 0; x < outBmp.Width; x++)
                    {
                        int i = (sy * Width + x / magnify) * 4;
                        row[x * 4 + 0] = Pixels[i + 2];   // B
                        row[x * 4 + 1] = Pixels[i + 1];   // G
                        row[x * 4 + 2] = Pixels[i + 0];   // R
                        row[x * 4 + 3] = Pixels[i + 3];   // A
                    }
                }
            }
        }
        finally
        {
            outBmp.UnlockBits(data);
        }
        return outBmp;
    }
}
